fix: calculate values of useHeaderHeight natively

tboba opened the pull request:
## Description

Currently, `useHeaderHeight()` hook returns default values, hard-coded in JS code. It uses `react-native-safe-area-context` under the hood to calculate top inset of the status bar and returns a header height `56` (for android) or `64` (for iOS). 
This PR introduces new `useAnimatedHeaderHeight` hook that calculates header height on each layout change.

## Changes

- Added support for calculating header height dynamically on the old / new architecture for Android and iOS.
- Added event that is being executed on header height change.
- Added new `useAnimatedHeaderHeight` hook for getting header height that is being calculated dynamically.

## Screenshots / GIFs

### Before

https://github.com/software-mansion/react-native-screens/assets/23281839/1e129ce8-cc62-4333-a6d8-f7283f018441

### After

https://github.com/software-mansion/react-native-screens/assets/23281839/5c11f371-c3f7-4eda-9969-4eb02f9de26a

## Test code and steps to reproduce

See `Test1802` in `TestsExample` or `FabricTestExample`.
There's also `Test1802b` included. It's a reproduction for an issue #1671 which shows that issue should be fixed after this change.

## Checklist

- [X] Included code example that can be used to test this change
- [X] Updated TS types
- [X] Updated documentation:
  - [X] https://github.com/software-mansion/react-native-screens/blob/main/guides/GUIDE_FOR_LIBRARY_AUTHORS.md
  - [X] https://github.com/software-mansion/react-native-screens/blob/main/native-stack/README.md
  - [X] https://github.com/software-mansion/react-native-screens/blob/main/src/types.tsx
- [x] Ensured that CI passes

hirbod commented:
Will this also mitigate the issue with the jumping header on Android when using tabs?

tboba commented:
Hi @hirbod, I think it should, but as this PR is still in an early state I cannot guarantee that. If this problem is strictly related to the value that is represented by the `useHeaderHeight()`, I'll try to find the solution for this bug and test it in this PR additionally 🎉

outaTiME commented:
That’s such good news @tboba, this dynamic calculation of the header size works also correctly on iOS modals?

tboba commented:
@outaTiME I looked onto that and it looks like header height is not being calculated on full screen modal - I already know why! I'll make a patch tomorrow for that 👍

tboba commented:
Hi @hirbod! Unfortunately, this PR still doesn't resolve that problem 😕 However, I'll try to find the solution for it after merging that one 👍

kkafar commented on ios/utils/RNSScreenModal.h:
```diff
@@ -0,0 +1,11 @@
+
+@interface RNSScreenModal : NSObject
+
++ (BOOL)isFullscreenModal:(UIModalPresentationStyle)presentationStyle;
++ (BOOL)isGrabbableModal:(UIModalPresentationStyle)presentationStyle;
++ (BOOL)isTransparentModal:(UIModalPresentationStyle)presentationStyle;
```
I would put these directly on `RNSScreenView` or maybe create a `ModalUtil` category or sth. No need to create separate interface in my opinion.

What do you think?

tboba replied:
I wouldn't go with category - I've tried that once and it was a bad idea. If I choose to move to the category, I would have to move the whole implementation there to omit circular imports. There I could use methods like `viewDidLayoutSubviews` to use newly-implemented functions there, but I found a strange behavior of modals while doing it - when I was trying to add `viewDidDisappear` to category, after closing the modal it was reappearing again 🤯 after closing it twice, then it disappeared completely.

I'll move them to RNSScreenView 😄

kkafar commented on src/native-stack/views/NativeStackView.tsx:
```diff
@@ -73,7 +78,7 @@ const MaybeNestedStack = ({
   const isHeaderInModal = isAndroid
     ? false
-    : stackPresentation !== 'push' && headerShown === true;
+    : stackPresentation !== 'push' && headerShown;
```
```suggestion
    : stackPresentation !== 'push' && headerShown === true;
```

It's JavaScript, let's be explicit.

kkafar reviewed (changes requested):
Great job 🥳 

I have not tested the usage of it yet. Just looked at the code for now.

General notes:

1. We should reconsider approach of setting the `dynamic` flag globally in screens, as I do not like having a possibly invalid value in `HeaderHeightContext` (do I understand it right? When `dynamic = false` the `.height` will have invalid value)
2. What do you think of `recalculate` -> `resolve` / `update` / just `calculate` (in all cases)?

I've also left some code notes below -- please answer them.

tboba commented on android/src/main/java/com/swmansion/rnscreens/Screen.kt:
```diff
@@ -236,6 +242,34 @@ class Screen constructor(context: ReactContext?) : FabricEnabledViewGroup(contex
+            else -> getSummarizedHeaderHeight()
+        }
+
+        UIManagerHelper.getEventDispatcherForReactTag(context as ReactContext, id)
+            ?.dispatchEvent(HeaderHeightChangeEvent(id, summarizedHeight))
```
Sounds good!

tboba replied:
@kkafar We've talked already about leaving it as it is for now, but as I was talking with Wojciech we made a decision to try to catch this value natively. RNSScreen should be sufficient for catching this value for now. I'll also try to reduce calls for recalculating header height value, as NativeStackView should already store the information about the value in previous screen. I'll try to investigate that 👍

tboba commented on android/src/main/java/com/swmansion/rnscreens/Screen.kt:
```diff
@@ -255,4 +289,11 @@ class Screen constructor(context: ReactContext?) : FabricEnabledViewGroup(contex
+    private fun Int.convertToDp(context: Context): Int {
+        val resources = context.resources
+        val metrics = resources.displayMetrics
+        val dp = this / (metrics.densityDpi / 160f)
```
In material design, one dp is equal to one pixel on a screen of density 160 dpi, so to calculate correct value I'm dividing density by 160 😅

kkafar replied:
Let's put this information as a comment into the code

WoLewicki commented on ios/RNSScreen.mm:
```diff
@@ -964,6 +989,73 @@ - (void)viewDidLayoutSubviews
+  [super dismissViewControllerAnimated:flag completion:completion];
+  BOOL isPresentedAsNativeModal = self.screenView.isPresentedAsNativeModal;
+
+  [self recalculateHeaderHeightIsModal:isPresentedAsNativeModal];
```
I think we could move the `self.screenView.isPresentedAsNativeModal` here directly since it is quite descriptive by itself

WoLewicki commented on ios/RNSScreen.mm:
```diff
@@ -964,6 +989,73 @@ - (void)viewDidLayoutSubviews
+  CGFloat navbarHeight = self.navigationController.navigationBar.frame.size.height;
+
+  // In case where screen is a modal, we want to calculate it's childViewController's height
+  if (isModal && self.childViewControllers.count > 0 && self.childViewControllers[0] != nil) {
```
I think something like this would be better since you cast it later.

```suggestion
  if (isModal && self.childViewControllers.count > 0 && [self.childViewControllers[0] isKindOfClass:UINavigationController.class]) {
```

WoLewicki replied:
Also, what if we had double nested stack? Will it work correctly in such case?

tboba replied:
@WoLewicki I checked that and it seems to work already!

WoLewicki commented on src/native-stack/views/NativeStackView.tsx:
```diff
@@ -304,6 +307,15 @@ const RouteView = ({
+          target: route.key,
+        });
+
+        setHeaderHeight(e.nativeEvent.newHeight);
```
Won't it cause many many rerenders of the whole `Screen` each time the height changes and even if the user did not subscribe to those changes?

tboba replied:
Hmm, that's a good catch 🤔 I'll reconsider it!

WoLewicki reviewed (commented):
I left some comments. Good job :tada:

kacperkapusciak commented on android/src/main/java/com/swmansion/rnscreens/ScreenViewManager.kt:
```diff
@@ -166,24 +167,16 @@ class ScreenViewManager : ViewGroupManager<Screen>(), RNSScreenManagerInterface<
-            MapBuilder.of("registrationName", "onHeaderBackButtonClicked"),
-            ScreenTransitionProgressEvent.EVENT_NAME,
-            MapBuilder.of("registrationName", "onTransitionProgress")
+    override fun getExportedCustomDirectEventTypeConstants(): MutableMap<String, Any> = mutableMapOf(
+            ScreenDismissedEvent.EVENT_NAME to MapBuilder.of("registrationName", "onDismissed"),
+            ScreenWillAppearEvent.EVENT_NAME to MapBuilder.of("registrationName", "onWillAppear"),
+            ScreenAppearEvent.EVENT_NAME to MapBuilder.of("registrationName", "onAppear"),
+            ScreenWillDisappearEvent.EVENT_NAME to MapBuilder.of("registrationName", "onWillDisappear"),
+            ScreenDisappearEvent.EVENT_NAME to MapBuilder.of("registrationName", "onDisappear"),
+            HeaderHeightChangeEvent.EVENT_NAME to MapBuilder.of("registrationName", "onHeaderHeightChange"),
+            HeaderBackButtonClickedEvent.EVENT_NAME to MapBuilder.of("registrationName", "onHeaderBackButtonClicked"),
+            ScreenTransitionProgressEvent.EVENT_NAME to MapBuilder.of("registrationName", "onTransitionProgress")
```
We have a Android linter setup in RNScreens repo. You can use it by running `yarn format-android`. It ensures consistency in the codebase and shouldn't be altered by hand.

tboba replied:
Sure, thanks for the tip!

kacperkapusciak commented on android/src/main/java/com/swmansion/rnscreens/Screen.kt:
```diff
@@ -236,6 +242,34 @@ class Screen constructor(context: ReactContext?) : FabricEnabledViewGroup(contex
 
+    fun recalculateHeaderHeight() {
+        val summarizedHeight = when (stackPresentation) {
+            StackPresentation.TRANSPARENT_MODAL -> 0
```
I believe transparent modals both on Android and iOS have headers and yielding `0` isn't correct

kacperkapusciak commented on android/src/main/java/com/swmansion/rnscreens/Screen.kt:
```diff
@@ -236,6 +242,34 @@ class Screen constructor(context: ReactContext?) : FabricEnabledViewGroup(contex
+            ?.dispatchEvent(HeaderHeightChangeEvent(id, summarizedHeight))
+    }
+
+    private fun getSummarizedHeaderHeight(): Int {
```
👍 to what's @kkafar saying about naming and let's also try to use similar naming on both platforms

kacperkapusciak commented on ios/RNSScreen.mm:
```diff
@@ -964,6 +989,73 @@ - (void)viewDidLayoutSubviews
+- (void)recalculateHeaderHeightIsModal:(BOOL)isModal
+{
+  if ([RNSScreenModal isTransparentModal:self.modalPresentationStyle]) {
+    [self.screenView notifyHeaderHeightChange:0];
```
I believe transparent modals both on Android and iOS have headers and yielding `0` isn't correct

WoLewicki replied:
Modals on `iOS` do not have headers, but we discussed it with @tboba already. On `Android`, modal is exactly the same fragment as the normal screen, but with the transparent background iirc. Or am I wrong @kkafar ?

kkafar commented on ios/RNSScreen.mm:
```diff
@@ -964,6 +1005,83 @@ - (void)viewDidLayoutSubviews
+
+  CGSize fallbackStatusBarSize = [[UIApplication sharedApplication] statusBarFrame].size;
+
+  if (@available(iOS 13.0, *)) {
+    CGSize primaryStatusBarSize = self.view.window.windowScene.statusBarManager.statusBarFrame.size;
+    if (primaryStatusBarSize.height == 0 || primaryStatusBarSize.width == 0)
+      return fallbackStatusBarSize;
+
+    return primaryStatusBarSize;
+  } else {
```
This code should be put inside ifdefs, as for someone on really old xcode it would fail to build. See [here](https://github.com/software-mansion/react-native-screens/blob/d40dc028a6e69a9aefbe56f6098fa9f517fc8190/ios/RNSScreen.mm#L120) for example.

tboba replied:
Sure! I was thinking why we're putting those ifdefs 🤔

kkafar replied:
FYI: The ```if (@available(iOS ... ))``` check is evaluated at runtime. When someone uses older xcode version without support for sdk >= iOS 13 the code inside the block could fail to build because of missing symbols at compile time.

tboba replied:
Introduced in https://github.com/software-mansion/react-native-screens/pull/1802/commits/1f5d586aff8373e79fa535af38756bd4c6da8e68.

kkafar commented on ios/RNSScreenStack.mm:
```diff
@@ -278,6 +310,11 @@ - (void)maybeAddToParentAndUpdateContainer
   }
 }
 
+- (void)addSubview:(UIView *)view
+{
+  [super addSubview:view];
+}
```
I believe it is redundant

```suggestion
```

tboba replied:
Introduced in https://github.com/software-mansion/react-native-screens/pull/1802/commits/9b2e310797d78f7c83296052878f00b276a4dfcd.

kkafar commented on ios/RNSScreen.h:
```diff
@@ -114,6 +118,9 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)notifyTransitionProgress:(double)progress closing:(BOOL)closing goingForward:(BOOL)goingForward;
 - (void)notifyDismissCancelledWithDismissCount:(int)dismissCount;
 - (BOOL)isModal;
+- (BOOL)isPresentedAsNativeModal;
+- (BOOL)isFullscreenModal;
+- (BOOL)isTransparentModal;
```
If these methods are used in `RNSScreen.mm` only and I believe they are -- we do not need to declare them in `@interface` -- these can be declared & defined entirely in `RNSScreen.mm`.

tboba replied:
Good point! I'll remove `isFullscreenModal` and `isTransparentModal`, as `isPresentedAsNativeModal` is being used in RNSScreenStack 😁

tboba replied:
Introduced in https://github.com/software-mansion/react-native-screens/pull/1802/commits/ae03a777dff550b45fe0a31055e154fef7ea49ca.

kkafar reviewed (changes requested):
Few more remarks, please answer them. Great job so far 🥳 
Also I have few points on whether we need to calculate header height on view controller dismissal (talking about iOS here)?

tboba commented on android/src/main/java/com/swmansion/rnscreens/Screen.kt:
```diff
@@ -255,4 +281,12 @@ class Screen constructor(context: ReactContext?) : FabricEnabledViewGroup(contex
         ORIENTATION, COLOR, STYLE, TRANSLUCENT, HIDDEN, ANIMATED, NAVIGATION_BAR_COLOR, NAVIGATION_BAR_HIDDEN
     }
+
+    // One dp is equal to one pixel on a screen of density 160 dpi.
+    private fun Int.convertToDp(context: Context): Int {
+        val resources = context.resources
+        val metrics = resources.displayMetrics
+        val dp = this / (metrics.densityDpi / 160f)
```
@kkafar Looks like it is! I've tested the `toDIPFromPixel` method from it and I'm getting the same values. Thanks for pointing this out.

tboba replied:
Introduced in https://github.com/software-mansion/react-native-screens/pull/1802/commits/78a81522b4970c9110ebfeb914cdbcf4b8b23a17.

tboba commented on TestsExample/src/Test1671.tsx:
```diff
@@ -0,0 +1,71 @@
+import React from 'react';
```
Introduced in `d1175c4`.

tboba commented on ios/RNSScreen.mm:
```diff
@@ -964,6 +1005,83 @@ - (void)viewDidLayoutSubviews
+    if ([self.childViewControllers[i] isKindOfClass:[RNScreensNavigationController class]]) {
+      return true;
+    }
+  }
```
Introduced in https://github.com/software-mansion/react-native-screens/pull/1802/commits/1f5d586aff8373e79fa535af38756bd4c6da8e68

tboba commented on ios/RNSScreen.mm:
```diff
@@ -964,6 +1005,83 @@ - (void)viewDidLayoutSubviews
+    }
+  }
+
+  return false;
```
Introduced in https://github.com/software-mansion/react-native-screens/pull/1802/commits/1f5d586aff8373e79fa535af38756bd4c6da8e68.

kkafar reviewed (commented):
We comin' home

tboba commented on TVOSExample/ios/TVOSExample.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>IDEDidComputeMac32BitWarning</key>
+	<true/>
+</dict>
+</plist>
```
It's surely unwanted 🤔

tboba replied:
Resolved in https://github.com/software-mansion/react-native-screens/pull/1802/commits/a68658b97f33653c31d281f10fda70130173b1e5.

tboba commented on native-stack/README.md:
````diff
@@ -750,10 +750,18 @@ navigation.popToTop();
 ### Measuring header's height
 
-To measure header's height, you can use `useHeaderHeight` hook.
+To measure header's height, you can use the `useHeaderHeight` or `useAnimatedHeaderHeight` hook.
+- `useHeaderHeight` calculates the header's height statically. The value provided by this hook only changes when the view appears or when there is a view rotation. 
+Use this hook if you are certain that your header height will not change, or if your component is heavy, and you want to avoid recalculating the header height with every view layout.
+- `useAnimatedHeaderHeight` calculates the header's height dynamically. The value provided by this hook changes with every view layout (such as when dragging a large header). 
+Please beware of using this hook in heavy components, as it may result in performance issues.
 
 ```tsx
+// for using useHeaderHeight
 import {useHeaderHeight} from 'react-native-screens/native-stack';
+
+// for using useAnimatedHeaderHeight
+import {useAnimatedHeaderHeight} from 'react-native-screens/native-stack';
 ```
````
I think it should! I've forgot to add a mention about it :eyes:

tboba replied:
Resolved in https://github.com/software-mansion/react-native-screens/pull/1802/commits/a68658b97f33653c31d281f10fda70130173b1e5.

kkafar commented on .gitignore:
```diff
@@ -21,6 +21,7 @@ DerivedData
 *.ipa
 *.xcuserstate
 project.xcworkspace
+*.plist
```
Hey, this should not be in top level `.gitignore`. See [`Info.plist`](https://github.com/software-mansion/react-native-screens/blob/main/FabricTestExample/ios/FabricTestExample/Info.plist) file which contains application specific configuration we want to have checked into version control.

tboba replied:
You're right, I can also see there's a couple of `IDEWorkspaceChecks.plist` folks already in the project. Removing this line 👍 https://github.com/software-mansion/react-native-screens/pull/1802/commits/6b23ba464e6cad2b4de6336fffc7f0eb35d6738b

kkafar reviewed (changes requested):
Approving besides what's mentioned below

kkafar reviewed (approved):
Great job :+1:

Green from me, but let's wait for others to go through the changes as we chatted today.

kacperkapusciak reviewed (approved):
Amazing work 🎉 

Please address the comments that I've left and also please check out whether this won't crash on a separate project without Reanimated installed.

Overall - good work! :shipit:

WoLewicki commented on ios/event/RNSHeaderHeightChangeEvent.mm:
```diff
@@ -0,0 +1,44 @@
+#import "RNSHeaderHeightChangeEvent.h"
```
Why is the folder named `event` not `events` as on android?

tboba replied:
That's a good point! Changed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/bb80aef75672c8f4aa9d13b956fe49e417be269d.

WoLewicki reviewed (approved):
If it works correctly in all scenarios, I think we are good to merge it :tada:

tboba commented on src/types.tsx:
```diff
@@ -72,6 +72,10 @@ export type HeaderSubviewTypes =
   | 'searchBar';
 
+export type HeaderChangeEventType = {
+  newHeight: number;
```
Sure! Fixed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/db7b50571e9c813e737b254294dd31459efb0e89.

tboba commented on native-stack/README.md:
```diff
@@ -750,10 +750,25 @@ navigation.popToTop();
 
-To measure header's height, you can use `useHeaderHeight` hook.
+To measure header's height, you can use the `useHeaderHeight`, `useAnimatedHeaderHeight` or `useReanimatedHeaderHeight` hook.
+- `useHeaderHeight` calculates the header's height statically. The value provided by this hook only changes when the view appears or when there is a view rotation. 
```
Fixed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/970dfcae90c7c8fdab9136a6e4f0bfc6aca7b738 🎉

tboba commented on native-stack/README.md:
```diff
@@ -750,10 +750,25 @@ navigation.popToTop();
-To measure header's height, you can use `useHeaderHeight` hook.
+To measure header's height, you can use the `useHeaderHeight`, `useAnimatedHeaderHeight` or `useReanimatedHeaderHeight` hook.
+- `useHeaderHeight` calculates the header's height statically. The value provided by this hook only changes when the view appears or when there is a view rotation. 
+Use this hook if you are certain that your header height will not change, or if your component is heavy, and you want to avoid recalculating the header height with every view layout.
```
Fixed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/970dfcae90c7c8fdab9136a6e4f0bfc6aca7b738 🎉

WoLewicki commented on ios/RNSScreenStack.mm:
```diff
@@ -57,6 +57,31 @@ - (UIViewController *)childViewControllerForStatusBarHidden
+
+    // If RNSScreen includes a navigation controller of type RNScreensNavigationController, it should not calculate
+    // header height, as it could have nested stack.
+    if (!(screenController.childViewControllers.count > 0 && screenController.hasAnyChildNavigators)) {
```
Why do we need to check the `count` here? Shouldn't `hasAnyChildNavigators` just return `NO` then?

tboba replied:
Seems like it should! I'll try tomorrow making an example with nested stack and check if `hasAnyChildNavigators` is sufficient there

tboba replied:
Can confirm that this part of if statement is unnecessary, I've changed it in https://github.com/software-mansion/react-native-screens/pull/1802/commits/271a4374fa394bf13d7516b348325698d8e86145.

WoLewicki commented on ios/RNSScreenStack.mm:
```diff
@@ -57,6 +57,31 @@ - (UIViewController *)childViewControllerForStatusBarHidden
 
+- (void)viewDidLayoutSubviews
+{
+  if ([self.topViewController isKindOfClass:[RNSScreen class]]) {
```
Shouldn't we call `[super viewDidLayoutSubviews]` here or is it not necessary?

tboba replied:
It's not necessary, everything works normally without the `super` (Xcode also doesn't complain about it) but I can add one 👍

tboba replied:
Added one in https://github.com/software-mansion/react-native-screens/pull/1802/commits/9557ec64ce6703d03bf0566d0b96fa2741b6ea18

WoLewicki commented on ios/RNSScreenStack.mm:
```diff
@@ -57,6 +57,31 @@ - (UIViewController *)childViewControllerForStatusBarHidden
+  if ([self.topViewController isKindOfClass:[RNSScreen class]]) {
+    RNSScreen *screenController = (RNSScreen *)self.topViewController;
+
+    // If RNSScreen includes a navigation controller of type RNScreensNavigationController, it should not calculate
+    // header height, as it could have nested stack.
```
What about both navigators having headers? I know it is a weird case but should we care about it?

tboba replied:
If you mean that we could have a nested stack with more than one header (which is a typical situation for nested stacks) then only the last one should call `calculateHeaderHeightIsModal`, so I assume we're right now caring about it (If I understand you correctly 😄)

WoLewicki commented on ios/RNSScreenStack.mm:
```diff
@@ -57,6 +57,31 @@ - (UIViewController *)childViewControllerForStatusBarHidden
+  [super dismissViewControllerAnimated:flag completion:completion];
+  BOOL wasModalPresented = ((RNSScreen *)self.presentedViewController).screenView.isModal;
+  // Check if presented modal was nested
+  BOOL isTopViewCtrNativeModal = ((RNSScreen *)self.topViewController).screenView.isPresentedAsNativeModal;
```
Is it ever `false`? I thought `dismissViewControllerAnimated` is called for modals only.

tboba replied:
It is - that's why we're checking if topViewController is _presented as native modal_. If it is, we're getting its top view controller's child view controller and passing its height.

Also, I'm trying to recall why there was `wasModalPresented` introduced 🤔 It seems that it is indeed obsolete.

tboba replied:
Removed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/94e52902d49a36e1f22a3059196898efc2a714ac, but I will check that further.

WoLewicki reviewed (commented):
Some more comments to answer when you have time 😅

tboba commented on FabricTestExample/src/Test1802.tsx:
```diff
@@ -0,0 +1,174 @@
+
+function ExampleScreen() {
+  const headerHeight = useAnimatedHeaderHeight();
+  // const headerHeight = useReanimatedHeaderHeight();
```
I tested it already and it was working without `useAnimatedStyles` 🤔

tboba commented on ios/RNSScreen.mm:
```diff
@@ -977,6 +1026,87 @@ - (void)viewDidLayoutSubviews
+
+// Checks whether this screen has any child view controllers of type RNScreensNavigationController.
+// Useful for checking if this screen has nested stack or is displayed at the top.
+- (BOOL)hasAnyChildNavigators
```
I was also thinking about that name, but I'm wondering if there could be a situation where the screen would not be nested and it still could have RNSNavigationController 🤔

WoLewicki replied:
There is no option for `RNSNavigationController` since it is the concept from our library for the stack navigator. If there was a child `RNSNavigationController` and it would not be a nested stack then it seems like a bug.

tboba replied:
Renamed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/d2622c4f0b80ab3939c09a3c369c3403b8d36138.

ha3 commented:
@tboba great work, thank you!

I just tested the new `useHeaderHeight` with a header with a search bar in iOS. Returning value doesn't include the height of the search bar. It would be great also to cover this case.

tboba commented:
Hi @ha3, thanks for letting us know! 🙏 

It looks like `use(Re)animatedHeaderHeight` calculates the header correctly (modifying the test a bit I can confirm this hook is returning correct value - transformY for Animated.Value puts FullWindowOverlay underneath the header).

<img width="540" alt="image" src="https://github.com/software-mansion/react-native-screens/assets/23281839/c60d4cc2-c512-4467-b013-5f0c013ce03d">

However, as I tested `useHeaderHeight` it seems that search bar is indeed not taken into account with calculations. 

<img width="475" alt="image" src="https://github.com/software-mansion/react-native-screens/assets/23281839/21f13049-d2f9-479a-b716-f3fe2247258f">

As we're now thinking about changing the implementation of `useHeaderHeight` completely, we will take care of it some day, but we won't fix this in this PR. Stay tuned 🙏

tboba commented on android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt:
```diff
@@ -103,6 +103,13 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
         mScreenFragments.removeAt(index)
+
+        // If screen at given index had another screen before in stack,
+        // calculate its header height.
```
We need this, because when there's a screen that is exiting to its parent it seems that it doesn't call `calculateHeaderHeight` (You can check it on video below, where I'm testing it in ScreensExample). However, it looks like this is not the case in TestsExample, so I assume this is necessary for the nested navigators, when we're traversing from child navigator to the parent one? 🤔 



Also, I think caring about triggering this method in other navigators is not necessary - shouldn't this method be called only for ScreenContainer that contains only screen with given index? Otherwise calls from lines 104-105 would throw an exception there because of mScreenFragments[index] 😄

WoLewicki replied:
I am not sure why should we trigger this event when going back to a screen? Could you elaborate on this more? I thought that each screen has its own header height measurement since each of them can have a header. So when going back to a screen, why would its header change from the time we left it?

tboba replied:
@WoLewicki I deeply tested that and it looks like currently NativeStackView loses the information about previous header height 😕 

https://github.com/software-mansion/react-native-screens/assets/23281839/40a388da-7cfa-475f-8ebf-d38506b10cb3

I assume that's because `Animated` does not hold the information between renders. I managed to solve that by wrapping `Animated` with useRef, but to be honest I'm not convinced if doing `React.useRef().current` and performing operations on it as usual is a good solution (and even practice?). What do you think? 🤔

tboba replied:
As I considered it, I don't see any points on which circumstances it might not work (I don't see any pitfalls either), so I assume we might go with that 🤔

tboba replied:
I've removed the android part of the code and changed the JS side here: https://github.com/software-mansion/react-native-screens/pull/1802/commits/2edab997027f41c73cc9422c085170bce290bcf2.

tboba commented:
/cc @kkafar @WoLewicki could you please take a quick look into those changes? I've removed 3 of 5 calls to `calculateHeaderHeight` that were recalculating unnecessarily header height on modal dismissals and changed a bit the structure of animatedHeaderHeight 🙇

WoLewicki commented on android/src/main/java/com/swmansion/rnscreens/ScreenContainer.kt:
```diff
@@ -103,6 +103,7 @@ open class ScreenContainer<T : ScreenFragment>(context: Context?) : ViewGroup(co
     open fun removeScreenAt(index: Int) {
         mScreenFragments[index].screen.container = null
         mScreenFragments.removeAt(index)
+
```
Looks like a leftover

tboba replied:
Resolved in https://github.com/software-mansion/react-native-screens/pull/1802/commits/246d3a555b18a3ab5416851ad86a92c1cd683deb.

WoLewicki commented on ios/RNSScreen.mm:
```diff
@@ -980,6 +1030,81 @@ - (void)viewDidLayoutSubviews
   }
 }
 
+// Checks whether this screen has any child view controllers of type RNScreensNavigationController.
```
Why the name here is `RNScreensNavigationController` and in code it is `RNSNavigationController`?

tboba replied:
Looks like the refactor tool didn't changed that when I was refactoring this class 🤔 Thanks for pointing this out!

tboba replied:
Resolved in https://github.com/software-mansion/react-native-screens/pull/1802/commits/fe776ca0bfc904f0376cc9958d7ef7661eb101cf.

tboba commented on ios/RNSScreen.mm:
```diff
@@ -980,6 +1030,81 @@ - (void)viewDidLayoutSubviews
+  BOOL isDraggableModalWithChildViewCtr =
+      isDraggableModal && self.childViewControllers.count > 0 && self.childViewControllers[0] != nil;
+
+  // When modal is floating (we can grab its header), we don't want to calculate status bar in it.
```
It looks like they have `headerShown` option enabled by default, but you're right that we should not count status bar here, as they're like 'transparent' 👍

tboba replied:
Resolved in https://github.com/software-mansion/react-native-screens/pull/1802/commits/a7d90a8da36c53fe6df20ecae7890a2598fb5363.

tboba commented on ios/RNSScreenStack.mm:
```diff
@@ -203,25 +218,25 @@ - (void)presentationControllerDidDismiss:(UIPresentationController *)presentatio
+      [self emitOnFinishTransitioningEvent];
+    });
   }
+#endif
```
Resolved in https://github.com/software-mansion/react-native-screens/pull/1802/commits/fe776ca0bfc904f0376cc9958d7ef7661eb101cf.

kkafar commented on ios/RNSScreen.mm:
```diff
@@ -980,6 +1031,81 @@ - (void)viewDidLayoutSubviews
+#endif // !TARGET_OS_TV
+}
+
+- (void)calculateHeaderHeightIsModal:(BOOL)isModal
+{
+  CGFloat navbarHeight = [self getCalculatedHeaderHeightIsModal:isModal];
+  CGSize statusBarSize = [self getCalculatedStatusBarHeightIsModal:isModal];
+
+  // Unfortunately, UIKit doesn't care about switching width and height options on screen rotation.
+  // We should check if user has rotated its screen, so we're choosing the minimum value between the
+  // width and height.
+  CGFloat statusBarHeight = MIN(statusBarSize.width, statusBarSize.height);
+  CGFloat totalHeight = navbarHeight + statusBarHeight;
+
+  [self.screenView notifyHeaderHeightChange:totalHeight];
+}
```
Do we want to dispatch event to JS on each call of this method? 
I would suggest splitting it into two separate methods:

1. actually resolving header height and returning it

```objective-c
- (CGFloat)calculateHeaderHeightIsModal:(BOOL)isModal
{ ... }
```
2. sending the event

Or in minimal version rename the method so that it is **clear** that it produces side-effects.

tboba replied:
Changed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/aabfc7edf48d5521de9d6e4b34a6fdceef8e0e85.

kkafar commented on ios/RNSScreenStack.mm:
```diff
@@ -61,6 +61,21 @@ - (UIViewController *)childViewControllerForStatusBarHidden
+    // Calculate header height during simple transition from one screen to another.
+    // If RNSScreen includes a navigation controller of type RNSNavigationController, it should not calculate
+    // header height, as it could have nested stack.
+    if (!screenController.hasNestedStack) {
```
`hasNestedStack` is a method, not property. Does this call actually work?

tboba replied:
As I investigated this method it looks like this method is being called right now (which is strange, indeed). It looks that even without this method it reports correct values, but this method is also reducing event calls to JS, so I would prefer to leave it here 🤞

kkafar commented on src/fabric/ScreenNativeComponent.ts:
```diff
@@ -21,6 +21,10 @@ type TransitionProgressEvent = Readonly<{
   goingForward: Int32;
 }>;
 
+type HeaderChangeEvent = Readonly<{
```
`HeaderHeightChange`?

tboba replied:
Renamed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/3c9c0fbe13cb07071e71254d249b2f90a964c3a0 👍

kkafar commented on src/types.tsx:
```diff
@@ -72,6 +72,10 @@ export type HeaderSubviewTypes =
   | 'center'
   | 'searchBar';
 
+export type HeaderChangeEventType = {
```
I would rename this type as well.

tboba replied:
Renamed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/3c9c0fbe13cb07071e71254d249b2f90a964c3a0 👍

kkafar reviewed (commented):
Few remarks and questions from me.

tboba commented on ios/RNSScreen.mm:
```diff
@@ -395,6 +396,29 @@ - (void)notifyDisappear
 
+- (void)notifyHeaderHeightChange:(double)headerHeight
+{
+#ifdef RCT_NEW_ARCH_ENABLED
+  if (_eventEmitter != nullptr) {
+    std::dynamic_pointer_cast<const facebook::react::RNSScreenEventEmitter>(_eventEmitter)
+        ->onHeaderHeightChange(
+            facebook::react::RNSScreenEventEmitter::OnHeaderHeightChange{.headerHeight = headerHeight});
+  }
+
+  RNSHeaderHeightChangeEvent *event =
+      [[RNSHeaderHeightChangeEvent alloc] initWithEventName:@"onHeaderHeightChange"
+                                                   reactTag:[NSNumber numberWithInt:self.tag]
+                                               headerHeight:headerHeight];
+  [[RCTBridge currentBridge].eventDispatcher sendEvent:event];
```
We've discussed that topic in the office 🤌

tboba commented on ios/RNSScreen.mm:
```diff
@@ -544,6 +563,34 @@ - (RNSScreenStackHeaderConfig *_Nullable)findHeaderConfig
+    case UIModalPresentationOverCurrentContext:
+      return true;
+    default:
+      return false;
```
Changed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/aabfc7edf48d5521de9d6e4b34a6fdceef8e0e85.

kkafar reviewed (approved):
I'm leaving green for the code changes, however I haven't tested the final version yet. I'll do it soon.

kkafar reviewed (approved):
🟢
